### DISCUSSION TITLE
Are there restrictions on object classes or conjoint attribute levels for reshape_projoint()?

### DISCUSSION THREAD

**Discussion by jcfoxe** [yhoriuchi/projoint#34] [Q&A] [answered]
Hello! Thanks for creating this awesome package! I am currently trying to reshape my data for subgroup analysis for a survey with 5 repeated choice tasks, 2 profiles, and 7 attributes/features, built using the Strezhnev tool. I particularly appreciate the package's capacity for subgroup analysis. 

Unfortunately, I cannot seem to pass the `projoint_data() `object created by `reshape_projoint()` to `projoint()` - I have used the `reshape_conjoint()` function in the `cjdata` package succesfully before when testing without my covariates, but I have had no luck passing the output of `reshape_projoint()` to `projoint()`, each time I get a message that the coefficients are "not defined because the design matrix is rank deficient". I have tried it both with and without my covariates and tried different settings but I can't seem to figure out what is wrong.  

I suppose this brings me to two questions,
1) **is there a way to transform the output of reshape_projoint() into an object with the class "dataframe" instead of an object of class "projoint_data"?** I'm not very familiar with S4 objects so I'm not sure if that's even feasible. 
2) **are there particular restrictions on the conjoint attributes, or indeed structures or classes the a variable needs to have in order for analysis to be done in projoint**? _(E.g., requiring objects to be as factors rather than characters, or how some packages don't support when attributes have different numbers of levels -- most of my conjoint attributes have four, except one attribute, which has 6 levels.)_ Just wondering what I may need to do before `reshape_projoint()` to get things to run.

**Comment by yhoriuchi** [comment, marked as answer]
Hi. I'm sorry for not getting back to you sooner. It's the end of the quarter at my institution, and I have been quite swamped in preparing for the final exam (tomorrow morning) and responding to student questions. Thank you so much for using this package. The comments and questions like yours are welcome to improve the quality of this package! 

1. You can just add `@estimates`. For example: `out <- projoint(cj_data)` and then `out_df <- out@estimates`. You can also save the attributes and levels in a data frame. For example, `cj_data <- reshape_projoint(data)` and then `attlevels <- cj_data@labels`.
2. I saw the same error message in my own application when I focused on very specific choice-level QoIs. But do you see the error message even when you use default settings? Do you have a reasonably large sample size? Can you try alternative estimates of standard errors (e.g., `.se_method = "simulation"`)?

**Reply by jcfoxe** [reply]
Thank you for your response, and no worries for the delay, it's a busy time! 

Initially, I wasn't able to use the addition of `@estimates` to yield `out <- projoint(cj_data)@estimates`. At first I still got the error message that each "coefficient not defined because the design matrix is rank deficient". However, after some correction, I managed to figure out that having the argument `.repeated = FALSE` within `reshape_projoint()` was the problem! I do have a repeated task, but misread the guidance on the arguments for the function, so this was 100% an error on my part, but it may be worth triple checking that this doesn't occur where there are multiple, varying tasks or when tasks do not repeat.  

```
cjdata <- reshape_projoint(
  .dataframe = d,
  .outcomes_ids = c("A", "B"),
  .alphabet = "F",
  .idvar = "ResponseId",
  .repeated = TRUE,
  .flipped = FALSE,
  .covariates = c("Q6_1", "Q6_2", "Q6_3", "Q6_4", "Q6_5", "Q6_6", "Q6_7", "Q7", "Q8", "Q9", "Q10", "Q11", "Q12", "Q13", "Q14", "Q15", "Q16", "Q17", "Q18", "Q19", "Q20", "Q21", "Q22", "Q23", "Q24", "Q25_1", "Q26"),
  .fill = FALSE, 
  .outcomes = c("F1", "F2", "F3", "F4", `"F5"))`
```

Then this - which works fine (as does the plot):
`out <- projoint(cjdata, .estimand = "mm")`

However, while I can now run my general analysis to obtain marginal means as above, the rank deficiency problem persists during subgroup analysis. I have used a dichotomous character variable (also tried converting to a factor, no difference) to try it out, and when I ran:

```
out2 <- projoint(cjdata,
                .by_var = "Q7",
                .estimand = "mm")
```
I was greeted with the same warning: "1 coefficient  not defined because the design matrix is rank deficient" - repeated for each coefficient. Have I misunderstood the restrictions on variables for subgroup analysis (dichotomous, as character only, per [instructions ](https://yhoriuchi.github.io/projoint/reference/projoint.html)for the `.by_var =` argument in `projoint()`)?

**Reply by yhoriuchi** [reply]
Hi. I am sorry for my late reply. The super busy teaching term ended, and I can now focus on this. I think your code should work. It's perhaps an issue specific to data. Do you have enough data for each subgroup? 

By the way, the backtick before "F5" and at the end of the line should be removed.

**Reply by yhoriuchi** [reply]
I actually noticed a similar issue for my own project. I will look into this problem.

**Reply by yhoriuchi** [reply]
In the process of overhauling this package, I believe that I have fixed this issue. If you still have any problem, please let me know.

**Comment by smonroe-byte** [comment]
Hello!
Thank you for making this awesome package! I am also having a similar issue as jcfoxe. When I try to get the corrected estimates for mm, I get the message that the coefficients are "not defined because the design matrix is rank deficient". When you get a chance, have you been able to look into this problem? Many thanks!

**Reply by yhoriuchi** [reply]
In the process of overhauling this package, I believe that I have fixed this issue. If you still have any problem, please let me know.

**Comment by smonroe-byte** [comment]
Thanks!
________________________________
